### ISSUE TITLE
firstindex(enumerate(...)) not defined

### ISSUE THREAD

**Issue by jamblejoe** [JuliaLang/julia#40704]
```
julia> firstindex(enumerate(1:3))
ERROR: MethodError: no method matching firstindex(::Base.Iterators.Enumerate{UnitRange{Int64}})
Closest candidates are:
  firstindex(::Any, ::Any) at abstractarray.jl:366
  firstindex(::Pair) at pair.jl:63
  firstindex(::AbstractChar) at char.jl:195
  ...
```

Shouldn't this be defined? The return of iterate(enumerate...) is a tuple of an index and the value of the original iterable. Shouldn't `firstindex` return the first index in that way?

I came across this because I wanted to parallelize a for loop with threads via
```
julia> for i in enumerate(2:5); @show i; end
i = (1, 2)
i = (2, 3)
i = (3, 4)
i = (4, 5)

julia> Threads.@threads for i in enumerate(2:5); @show i; end
ERROR: TaskFailedException
Stacktrace:
 [1] wait
   @ .\task.jl:317 [inlined]
 [2] threading_run(func::Function)
   @ Base.Threads .\threadingconstructs.jl:34
 [3] top-level scope
   @ .\threadingconstructs.jl:93

    nested task error: MethodError: no method matching firstindex(::Base.Iterators.Enumerate{UnitRange{Int64}})
    Closest candidates are:
      firstindex(::Any, ::Any) at abstractarray.jl:366
      firstindex(::Pair) at pair.jl:63
      firstindex(::AbstractChar) at char.jl:195
      ...
    Stacktrace:
     [1] (::var"#14#threadsfor_fun#7"{Base.Iterators.Enumerate{UnitRange{Int64}}})(onethread::Bool)
       @ Main .\threadingconstructs.jl:66
     [2] (::var"#14#threadsfor_fun#7"{Base.Iterators.Enumerate{UnitRange{Int64}}})()   
       @ Main .\threadingconstructs.jl:48

julia> 
``` 

```
julia> versioninfo()
Julia Version 1.6.0
Commit f9720dc2eb (2021-03-24 12:55 UTC)
Platform Info:
  OS: Windows (x86_64-w64-mingw32)
  CPU: Intel(R) Core(TM) i7-8665U CPU @ 1.90GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-11.0.1 (ORCJIT, skylake)
Environment:
  JULIA_EDITOR = "C:\Users\gnakerst\AppData\Local\atom\app-1.56.0\atom.exe"  -a
  JULIA_NUM_THREADS = 4
```

**Comment by jishnub** [comment]
It should be defined, since it has axes defined

```julia
julia> axes(enumerate(1:4))
(Base.OneTo(4),)
```

Evidently many types from `Base.Iterators` don't have this defined.

**Comment by Seelengrab** [comment]
`firstindex` implies `getindex` at a later point, which may not be possible for all iterators (since it may have to consume them to get there - not all iterables support random indexing after all). For example, `Iterators.filter` can't have random indexing without consuming the underlying iterator since you have to check each possible value before knowing whether or not the value is in the filtered collection or not. Since `enumerate` has to be able to wrap arbitrary iterables, it doesn't necessarily have `getindex` available on the underlying iterator (but there could be an argument made for adding it to `enumerate`, though that may be confusing when doing e.g. `enumerate(::String)`, since those indices don't correspond to the indices on the underlying string). 

The `axes` method that's hit is the fallback method, which only works because `enumerate` has a length.

**Comment by Qfl3x** [comment]
I'm interested in this issue (Hopefully my first issue), I'm not too good but decent with Julia. 

What's confusing me is why would anyone use `enumerate` on non-indexable iterators?

**Comment by Qfl3x** [comment]
@jamblejoe Okay, so after further investigation I think Iterators are off and issues like these will keep popping up. While I could easily hardcode it (Already have: `first(getindex(axes(A),1))`), such issues will keep popping up. If we implement `getindex` and `eachindex` to each Iterator (where it makes sense)  and make them all subtypes of `AbstractArray` properly. Will that solve the issues?

That will be a lot of work though, since it will imply implementing the primitive `AbstractArray` functions to each Iterator. It should be faster than `collect`ing still.

**Comment by simeonschaub** [comment]
There is no need to implement this for arbitrary iterators. `firstindex(enumerate(itr))` should simply return `firstindex(itr)`. For iterators that don't support `firstindex` it's fine for this to just throw a `MethodError`.

**Comment by Qfl3x** [comment]
Implemented it simply here

**Comment by adienes** [comment]
> firstindex(enumerate(itr)) should simply return firstindex(itr)

I disagree very strongly with this. it breaks all kinds of correspondences one might expect about the relationships between `first`, `firstindex`, `getindex`
